File: feature-venues-history/src/main/java/home/oleg/placesnearme/feature_venues_history/di/VenueHistoryComponent.java
package home.oleg.placesnearme.feature_venues_history.di;

import dagger.BindsInstance;
import dagger.Component;
import home.oleg.placesnearme.AppApiProvider;
import home.oleg.placesnearme.api.AppApi;
import home.oleg.placesnearme.feature_venues_history.VenuesHistoryFragment;
import home.oleg.placesnearme.scopes.FeatureScope;

@FeatureScope
@Component(dependencies = AppApi.class, modules = VenuesHistoryModule.class)
public interface VenueHistoryComponent {
    void inject(VenuesHistoryFragment target);

    @Component.Builder
    interface Builder {
        VenueHistoryComponent.Builder appComponent(AppApi appApi);

        @BindsInstance
        VenueHistoryComponent.Builder bind(VenuesHistoryFragment target);

        VenueHistoryComponent build();
    }

    final class Injector {
        private Injector() {
        }

        public static void inject(VenuesHistoryFragment fragment) {
            AppApi appApi = AppApiProvider.Initializer.getAppApi(fragment);
            DaggerVenueHistoryComponent.builder()
                    .appComponent(appApi)
                    .bind(fragment)
                    .build()
                    .inject(fragment);
        }
    }
}

File: feature-venues-history/src/main/java/home/oleg/placesnearme/feature_venues_history/di/VenueHistoryComponent.kt
package home.oleg.placesnearme.feature_venues_history.di

import dagger.BindsInstance
import dagger.Component
import home.oleg.placesnearme.coredata.di.RepoApi
import home.oleg.placesnearme.coredi.module.ViewModelFactoryModule
import home.oleg.placesnearme.coredi.scopes.FeatureScope
import home.oleg.placesnearme.corettools.di.ToolsApi
import home.oleg.placesnearme.feature_venues_history.presentation.ui.VenuesHistoryFragment

@FeatureScope
@Component(
        dependencies = [ToolsApi::class, RepoApi::class],
        modules = [ViewModelFactoryModule::class, VenuesHistoryModule::class])
interface VenueHistoryComponent {
    fun inject(target: VenuesHistoryFragment)

    @Component.Builder
    interface Builder {
        fun toolsProvider(toolsProvider: ToolsApi): VenueHistoryComponent.Builder
        fun repoComponent(repoApi: RepoApi): VenueHistoryComponent.Builder

        @BindsInstance
        fun bind(target: VenuesHistoryFragment): VenueHistoryComponent.Builder

        fun build(): VenueHistoryComponent
    }

    object Injector {

        fun inject(fragment: VenuesHistoryFragment) {
            val toolsProvider = ToolsApi.getInstance(fragment.activity!!.applicationContext)
            val repoApi = RepoApi.getInstance(fragment.activity!!.applicationContext)

            DaggerVenueHistoryComponent.builder()
                    .toolsProvider(toolsProvider)
                    .repoComponent(repoApi)
                    .bind(fragment)
                    .build()
                    .inject(fragment)
        }
    }
}
